content/better-fors.md

```diff
@@ -0,0 +1,381 @@
+---
+layout: sip
+permalink: /sips/:title.html
+stage: design
+status: submitted
+title: SIP-62 - For comprehension improvements
+---
+
+**By: Kacper Korban (VirtusLab)**
+
+## History
+
+| Date          | Version            |
+|---------------|--------------------|
+| June 6th 2023 | Initial Draft      |
+| Feb 15th 2024 | Reviewed Version   |
+
+## Summary
+
+`for`-comprehensions in Scala 3 improved their usability in comparison to Scala 2, but there are still some pain points relating both usability of `for`-comprehensions and simplicity of their desugaring.
+
+This SIP tries to address some of those problems, by changing the specification of `for`-comprehensions. From user perspective, the biggest change is allowing aliases at the start of the `for`-comprehensions. e.g.
+
+```
+for {
+  x = 1
+  y <- Some(2)
+} yield x + y
+```
+
+## Motivation
+
+There are some clear pain points related to Scala'3 `for`-comprehensions and those can be divided into two categories:
+
+1. User-facing and code simplicity problems
+
+    Specifically, for the following example written in a Haskell-style do-comprehension
+
+    ```haskell
+    do
+      a = largeExpr(arg)
+      b <- doSth(a)
+      combineM(a, b)
+    ```
+    in Scala we would have to write
+
+    ```scala
+    val a = largeExpr(b)
+    for
+      b <- doSth(a)
+      x <- combineM(a, b)
+      yield x
+    ```
+
```

Comment:
Elsewhere, we discussed how making the final `map` optional would be a breaking change. But what if we went the haskell route, and only removed the final `map` if the final `yield` is also removed?


    ```scala
    val a = largeExpr(b)
    for
      b <- doSth(a)
      combineM(a, b)
    ```

1. This is currently invalid syntax, so no backwards compat concerns
2. We get to remove the useless `map` from the generated code
3. We *also* get to remove the useless `yield` from the source code!
4. The correspondence between the `map` and `yield` is direct. No "magic" map-removal depending on the shape of the final expression
5. The final source code is shorter and more concise
6. The final source code looks more like other languages with similar syntax, e.g. Haskell's `do` or F#'s `return!` in computation expressions

Reply:
I did consider this change, but I didn't manage to get it to work in the parser. (I tried going in the [reflect-like direction](https://github.com/lampepfl/dotty/compare/main...dotty-staging:dotty:monadic-blocks#diff-ff13049a579bd41563e00e4391bcf7a18d9345b754683c44b2d3e971f3fd150a) to get better syntax for this case)
But if we were to be able to distinguish a pattern from an expression, then yet another problem is automatically solved -- we can get rid of the unnecessary `_ <-` with monadic 'unit' expressions.

One workaround from a [pre-SIP](https://contributors.scala-lang.org/t/pre-sip-improve-for-comprehensions-functionality/3509) is to have a `yield <- monadicExpr` syntax.

Reply:
would it be controversial to reuse `return` as in
```scala
for x <- List(1,2,3) return List(x, x)
```

Reply:
I would be against this for several reasons, including: 1) yet another syntax to learn, but worse: 2) my experience is that many learners have all sorts of connotations on the `return` keyword with all sorts of misconceptions about how it works in Scala so I think it is best to say "don't use return" and for some it takes a while not to end everything with return if they are used to that...